### PR TITLE
refactor(core): Use hyperswitch_domain_models within the Payments Core instead of api_models

### PR DESCRIPTION
**Pull request by prasunna09**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

- Use hyperswitch_domain_models within the Payments Core instead of api_models
- introduced a new payment method data  Network Token in domain models

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

TestCases:- 
Sanity testing for all the payments flows including save card flow and pm_auth flow.

stripe testing - 
Screenshot 2024-08-06 at 5 00 14 PM
 cybersource - 
Screenshot 2024-08-06 at 5 07 50 PM
saved card payment - 
Screenshot 2024-08-06 at 4 36 05 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/payments/helpers.rs [hunk 1]
```diff
@@ -1755,13 +1756,14 @@ pub async fn retrieve_payment_method_with_temporary_token(
...
-        Some(api::PaymentMethodData::Card(card)) => {
+        Some(domain::PaymentMethodData::Card(card)) => {
             let mut updated_card = card.clone();
             let mut is_card_updated = false;
 
             // The card_holder_name from locker retrieved card is considered if it is a non-empty string or else card_holder_name is picked
             // from payment_method_data.card_token object
-            let name_on_card = if let Some(name) = card.card_holder_name.clone() {
+            let name_on_card = if let Some(name) = card.nick_name.clone() {
+                //todo!
```
Comment (lsampras) [inline crates/router/src/core/payments/helpers.rs:1766]: this needed?

File: crates/router/src/core/payments/helpers.rs [hunk 2]
```diff
@@ -1770,7 +1772,7 @@ pub async fn retrieve_payment_method_with_temporary_token(
                         })
                         .or(Some(name))
                 } else {
-                    card.card_holder_name.clone()
+                    card.nick_name.clone() //todo!
```
Comment (lsampras) [inline crates/router/src/core/payments/helpers.rs:1775]: can we add details about what has to be done

**Review by lsampras — DISMISSED** [review]

**Review by SamraatBansal — DISMISSED** [review]
LGTM for connector changes

File: crates/router/src/core/payments/helpers.rs [hunk 3]
```diff
@@ -2028,7 +2030,7 @@ pub async fn make_pm_data<'a, F: Clone, R>(
     }
 
     // TODO: Handle case where payment method and token both are present in request properly.
-    let (payment_method, pm_id) = match (request, payment_data.token_data.as_ref()) {
```
Comment (Chethan-rao) [inline crates/router/src/core/payments/helpers.rs:2031]: I don't think this clone is required? u can match on reference of `request` itself

File: crates/router/src/core/payments/helpers.rs [hunk 4]
```diff
@@ -2059,7 +2061,7 @@ pub async fn make_pm_data<'a, F: Clone, R>(
 
         (Some(_), _) => {
             let (payment_method_data, payment_token) = payment_methods::retrieve_payment_method(
-                request,
+                &request.map(Into::into),
```
Comment (Chethan-rao) [inline crates/router/src/core/payments/helpers.rs:2064]: You can just pass `request` if u make this a reference above. No type conversion required as request type is already `Option<domain::PaymentMethodData>`

File: crates/router/src/core/payments/operations/payment_complete_authorize.rs [hunk 5]
```diff
@@ -314,7 +314,7 @@ impl<F: Send + Clone> GetTracker<F, PaymentData<F>, api::PaymentsRequest> for Co
             payment_method_data: request
                 .payment_method_data
                 .as_ref()
-                .and_then(|pmd| pmd.payment_method_data.clone()),
+                .and_then(|pmd| pmd.payment_method_data.clone().map(Into::into).clone()),
```
Comment (Chethan-rao) [inline crates/router/src/core/payments/operations/payment_complete_authorize.rs:317]:
```suggestion
                .and_then(|pmd| pmd.payment_method_data.clone().map(Into::into)),
```
Clone not required

File: crates/router/src/core/payments/operations/payment_confirm.rs [hunk 6]
```diff
@@ -1065,6 +1078,9 @@ impl<F: Clone> UpdateTracker<F, PaymentData<F>, api::PaymentsRequest> for Paymen
             .as_ref()
             .get_required_value("profile_id")
             .change_context(errors::ApiErrorResponse::InternalServerError)?;
+
+        // payment_data.payment_attempt.clone();
+
```
Comment (Chethan-rao) [inline crates/router/src/core/payments/operations/payment_confirm.rs:1083]: remove this

File: crates/router/src/core/payments/operations/payment_confirm.rs [hunk 7]
```diff
@@ -1117,14 +1133,19 @@ impl<F: Clone> UpdateTracker<F, PaymentData<F>, api::PaymentsRequest> for Paymen
         };
 
         let customer_details = payment_data.payment_intent.customer_details.clone();
-        let business_sub_label = payment_data.payment_attempt.business_sub_label.clone();
-        let authentication_type = payment_data.payment_attempt.authentication_type;
+        let business_sub_label = payment_data
+            .payment_attempt
+            .clone()
+            .business_sub_label
+            .clone();
```
Comment (Chethan-rao) [inline crates/router/src/core/payments/operations/payment_confirm.rs:1140]:
```suggestion
        let business_sub_label = payment_data
            .payment_attempt
            .business_sub_label
            .clone();
```
same here

File: crates/router/src/core/payments/operations/payment_confirm.rs [hunk 8]
```diff
@@ -1117,14 +1133,19 @@ impl<F: Clone> UpdateTracker<F, PaymentData<F>, api::PaymentsRequest> for Paymen
...
+        let authentication_type = payment_data.payment_attempt.clone().authentication_type;
```
Comment (Chethan-rao) [inline crates/router/src/core/payments/operations/payment_confirm.rs:1141]:
```suggestion
        let authentication_type = payment_data.payment_attempt.authentication_type;
```
same here

File: crates/router/src/core/payments/operations/payment_confirm.rs [hunk 9]
```diff
@@ -1117,14 +1133,19 @@ impl<F: Clone> UpdateTracker<F, PaymentData<F>, api::PaymentsRequest> for Paymen
...
 
         let (shipping_address_id, billing_address_id, payment_method_billing_address_id) = (
             payment_data.payment_intent.shipping_address_id.clone(),
             payment_data.payment_intent.billing_address_id.clone(),
             payment_data
                 .payment_attempt
+                .clone()
                 .payment_method_billing_address_id
                 .clone(),
```
Comment (Chethan-rao) [inline crates/router/src/core/payments/operations/payment_confirm.rs:1150]:
```suggestion
                .payment_attempt
                .payment_method_billing_address_id
                .clone(),
```
same here

File: crates/router/src/core/payments/operations/payment_create.rs [hunk 10]
```diff
@@ -846,13 +860,11 @@ impl PaymentCreate {
...
         let mut additional_pm_data = request
             .payment_method_data
-            .as_ref()
-            .and_then(|payment_method_data_request| {
-                payment_method_data_request.payment_method_data.as_ref()
-            })
-            .async_map(|payment_method_data| async {
+            .clone()
+            .and_then(|payment_method_data_request| payment_method_data_request.payment_method_data)
+            .async_and_then(|payment_method_data| async {
```
Comment (Chethan-rao) [inline crates/router/src/core/payments/operations/payment_create.rs:865]: No need to clone entire payment_method_data of request. Instead take `request.payment_method_data` as reference and clone `payment_method_data_request.payment_method_data` inside and_then

File: crates/router/src/types/storage/payment_method.rs [hunk 11]
```diff
@@ -10,7 +10,10 @@ pub use diesel_models::payment_method::{
     TokenizeCoreWorkflow,
 };
 
-use crate::types::api::{self, payments};
+use crate::types::{
+    api::{self},
+    domain::{self},
+};
```
Comment (Chethan-rao) [inline crates/router/src/types/storage/payment_method.rs:16]:
```suggestion
use crate::types::{
    api,domain
};
```

**Review by Chethan-rao — DISMISSED** [review]

**Review by Narayanbhat166 — DISMISSED** [review]

**Review by Narayanbhat166 — DISMISSED** [review]

**Review by Chethan-rao — APPROVED** [review]

**Review by Chethan-rao — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by lsampras — APPROVED** [review]

**Review by SamraatBansal — APPROVED** [review]

**Pull request merged**